guillotina_amqp/worker.py

```diff
@@ -34,103 +49,179 @@ def state_manager(self):
 
     @property
     def num_running(self):
+        """Returns the number of currently running jobs"""
         return len(self._running)
 
     async def handle_queued_job(self, channel, body, envelope, properties):
+        """Callback triggered when there is a new job in the job channel (e.g:
+        a new task in a rabbitmq queue)
+
+        """
+        # Deserialize job description
         if not isinstance(body, str):
             body = body.decode('utf-8')
         data = json.loads(body)
+
+        # Set status to scheduled
         task_id = data['task_id']
         dotted_name = data['func']
         await self.state_manager.update(task_id, {
             'status': 'scheduled'
         })
         logger.info(f'Received task: {task_id}: {dotted_name}')
+
+        # Block if we reached maximum number of running tasks
         while len(self._running) >= self._max_size:
             await asyncio.sleep(self.sleep_interval)
             self.last_activity = time.time()
+
+        # Create job object
         self.last_activity = time.time()
         job = Job(self.request, data, channel, envelope)
-
+        # Get the redis lock on the task so no other worker takes it
         _id = job.data['task_id']
         ts = TaskState(_id)
-        res = await ts.acquire(timeout=120)
-        if not res:
-            raise Exception()
 
+        if await ts.is_canceled():
+            logger.warning(f'Task {_id} has already been canceled')
+            raise TaskAlreadyCancelled(_id)
+        try:
+            res = await ts.acquire()
+        except TaskAlreadyAcquired:
+            logger.warning(f'Task {_id} is already running in another worker')
+            raise
 
+        # Record job's data into global state
+        await self.state_manager.update(task_id, {
+            'job_data': job.data,
+        })
+
+        # Add the task to the loop and start it
         task = self.loop.create_task(job())
         task._job = job
         job.task = task
         self._running.append(task)
         task.add_done_callback(self._done_callback)
 
     def _done_callback(self, task):
+        """This is called when a job finishes execution"""
         self._done.append(task)
         self.total_run += 1
 
     async def start(self):
+        """Called on worker startup. Connects to the rabbitmq. Declares and
+        configures the different queues.
+
+        """
         channel, transport, protocol = await amqp.get_connection()
 
+        EXCHANGE = app_settings['amqp']['exchange']
```

Comment:
@gitcarbs can you check this part of the code to confirm that the queues config is OK according to the process we discussed?